### PR TITLE
HPCC-23367 Thor-on-demand - 1st cut.

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - please review

File: common/workunit/workunit.cpp [hunk 1]
```diff
@@ -13182,13 +13182,51 @@ static void setResources(StringBuffer &jobYaml, const IConstWorkUnit *workunit,
...
+}
+
+void launchK8sJob(const char *componentName, const char *wuid, const char *job, const std::list<std::pair<std::string, std::string>> &extraParams)
+{
+    VStringBuffer jobname("%s-%s", componentName, job);
+    jobname.toLowerCase();
+    VStringBuffer args("\"--workunit=%s\"", wuid);
+    for (const auto &p: extraParams)
+        args.append(',').newline().append("\"--").append(p.first.c_str()).append('=').append(p.second.c_str()).append("\"");
+    VStringBuffer jobSpecFilename("/etc/config/%s.yaml", componentName); // default to /etc/config/jobspec.yaml
```
Comment (richardkchapman) [inline common/workunit/workunit.cpp]: Does it default to that?

File: dockerfiles/hpcc/templates/eclagent.yaml [hunk 2]
```diff
@@ -29,7 +29,7 @@ spec:
 {{ include "hpcc.addVolumes" $ | indent 6 }}
 ---
 {{- include "hpcc.generateConfigMap" (dict "root" $ "component" "eclagent" "me" .) }}
-  jobspec.yaml: |
+  eclagent.yaml: |
```
Comment (richardkchapman) [inline dockerfiles/hpcc/templates/eclagent.yaml]: not wild about the name - eclagent-job or something perhaps?

File: dockerfiles/hpcc/templates/eclccserver.yaml [hunk 3]
```diff
@@ -31,7 +31,7 @@ spec:
 {{ include "hpcc.addVolumes" $ | indent 6 }}
 ---
 {{- include "hpcc.generateConfigMap" (dict "root" $ "component" "eclccserver" "me" .) }}
-  jobspec.yaml: |
+  eclccserver.yaml: |
```
Comment (richardkchapman) [inline dockerfiles/hpcc/templates/eclccserver.yaml]: see above

File: dockerfiles/hpcc/templates/thor.yaml [hunk 4]
```diff
@@ -1,84 +1,137 @@
...
+      - name: {{ $eclagentName | quote }}
         ports:
           - containerPort: {{ $masterPort }}
         args: [
                 {{ include "hpcc.configArg" . }},
-                {{ include "hpcc.daliArg" $ }},
-                --masterport={{ $masterPort }},
-                --numSlaves={{ $thor.numSlaves }}
+                {{ include "hpcc.daliArg" $ }}
+                {{- if $thor.maxActive }},
```
Comment (richardkchapman) [inline dockerfiles/hpcc/templates/thor.yaml]: Shouldn't need things like this now we pass all settings through via config file

File: dockerfiles/hpcc/templates/thor.yaml [hunk 5]
```diff
@@ -1,84 +1,137 @@
...
-  replicas: {{ $thor.numSlaves }}
-  selector:
-    matchLabels:
-      run: {{ $slaveName | quote }}
-  template:
+{{- include "hpcc.generateConfigMap" (dict "root" $ "component" "thor" "me" .) }}
+{{- if $thor.eclagent }}
+    eclagent:
+{{ toYaml $thor.eclagent | indent 6 }}
+      name: myeclagent-thor
```
Comment (richardkchapman) [inline dockerfiles/hpcc/templates/thor.yaml]: this looks too hardwired...

File: dockerfiles/hpcc/templates/thor.yaml [hunk 6]
```diff
@@ -1,84 +1,137 @@
...
+      template:
+        spec:
+          serviceAccountName: hpcc
+          {{- include "hpcc.checkDataStorageHostMount" (dict "root" $) | indent 10 }}
+          containers:
+          - name: %jobname
+            image: {{ include "hpcc.imageName" (dict "root" $ "me" . "imagename" "eclagent") }}
+            resources:
+              requests:
+                #request-memory
```
Comment (richardkchapman) [inline dockerfiles/hpcc/templates/thor.yaml]: was there supposed to be something here?

File: thorlcr/master/thmastermain.cpp [hunk 7]
```diff
@@ -557,12 +557,17 @@ version: 1.0
   daliServers: dali
   watchdogEnabled: true
   watchdogProgressEnabled: true
+  cost:
+    thor:
+      master: "0.000002"
```
Comment (richardkchapman) [inline thorlcr/master/thmastermain.cpp]: this seems unrelated

**Review by richardkchapman — COMMENTED** [review]
Some minor comments

File: dockerfiles/hpcc/templates/thor.yaml (hunk 5 above)
Comment (jakesmith) [reply dockerfiles/hpcc/templates/thor.yaml]: should be there, was an experiment to work with matching environment.xml and differentiate between hthor and thor cluster as it is, but replaced by specifying queueNames.
Will remove.

File: thorlcr/master/thmastermain.cpp (hunk 7 above)
Comment (jakesmith) [reply thorlcr/master/thmastermain.cpp]: it's defaults from Shamer's stuff that was in the now deleted thor.xml

File: common/workunit/workunit.cpp (hunk 1 above)
Comment (jakesmith) [reply common/workunit/workunit.cpp]: no, comment is outdated, should be removed.

File: dockerfiles/hpcc/templates/eclagent.yaml (hunk 2 above)
Comment (jakesmith) [reply dockerfiles/hpcc/templates/eclagent.yaml]: agree, will change these.

File: dockerfiles/hpcc/templates/eclccserver.yaml (hunk 3 above)
Comment (jakesmith) [reply dockerfiles/hpcc/templates/eclccserver.yaml]: will change

File: dockerfiles/hpcc/templates/thor.yaml (hunk 4 above)
Comment (jakesmith) [reply dockerfiles/hpcc/templates/thor.yaml]: right.. I think was under a different section at some point. will remove.

File: dockerfiles/hpcc/templates/thor.yaml (hunk 6 above)
Comment (jakesmith) [reply dockerfiles/hpcc/templates/thor.yaml]: hm,  it's the same as eclagent.yaml - should it not be?

Comment (richardkchapman) [reply dockerfiles/hpcc/templates/thor.yaml]: I think I had just forgotten how it works...

It gets substituted in from wu values - which you may not want for the eclagent associated with a thor (the wu values probably shoudl go into the thor itself instead)

File: dockerfiles/hpcc/templates/eclagent.yaml (hunk 2 above)
Comment (jakesmith) [reply dockerfiles/hpcc/templates/eclagent.yaml]: however.. this is found via the component name at the moment.
so e.g. :
                runK8sJob("eclagent", wuid, wuid);

runK8Job uses componentName to find resources to set from the workunit in setResources.

Could be jobspec name, and component/type name..

File: dockerfiles/hpcc/templates/eclccserver.yaml [hunk 8]
```diff
@@ -31,7 +31,7 @@ spec:
 {{ include "hpcc.addVolumes" $ | indent 6 }}
 ---
 {{- include "hpcc.generateConfigMap" (dict "root" $ "component" "eclccserver" "me" .) }}
-  eclccserver.yaml: |
+  eclccserver-job.yaml: |
```
Comment (richardkchapman) [inline dockerfiles/hpcc/templates/eclccserver.yaml]: not jobspec?

Comment (jakesmith) [reply dockerfiles/hpcc/templates/eclccserver.yaml]: yep, fixed

**Comment by jakesmith** [comment]
@richardkchapman - please see review changes

**Comment by jakesmith** [comment]
@richardkchapman - squashed
